rules/terraform_no_short_circuit_evaluation.go

```diff
@@ -0,0 +1,139 @@
+package rules
+
+import (
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/terraform-linters/tflint-plugin-sdk/tflint"
+	"github.com/terraform-linters/tflint-ruleset-terraform/project"
+)
+
+// TerraformNoShortCircuitEvaluationRule checks for attempts to use short-circuit evaluation
+type TerraformNoShortCircuitEvaluationRule struct {
+	tflint.DefaultRule
+}
+
+// NewTerraformNoShortCircuitEvaluationRule returns a new rule
+func NewTerraformNoShortCircuitEvaluationRule() *TerraformNoShortCircuitEvaluationRule {
+	return &TerraformNoShortCircuitEvaluationRule{}
+}
+
+// Name returns the rule name
+func (r *TerraformNoShortCircuitEvaluationRule) Name() string {
+	return "terraform_no_short_circuit_evaluation"
+}
+
+// Enabled returns whether the rule is enabled by default
+func (r *TerraformNoShortCircuitEvaluationRule) Enabled() bool {
+	return true
+}
+
+// Severity returns the rule severity
+func (r *TerraformNoShortCircuitEvaluationRule) Severity() tflint.Severity {
+	return tflint.WARNING
+}
+
+// Link returns the rule reference link
+func (r *TerraformNoShortCircuitEvaluationRule) Link() string {
+	return project.ReferenceLink(r.Name())
+}
+
+// Check checks for attempts to use short-circuit evaluation
+func (r *TerraformNoShortCircuitEvaluationRule) Check(runner tflint.Runner) error {
+	path, err := runner.GetModulePath()
+	if err != nil {
+		return err
+	}
+	if !path.IsRoot() {
+		// This rule does not evaluate child modules
+		return nil
+	}
+
+	diags := runner.WalkExpressions(tflint.ExprWalkFunc(func(expr hcl.Expression) hcl.Diagnostics {
+		if binaryOpExpr, ok := expr.(*hclsyntax.BinaryOpExpr); ok {
+			if binaryOpExpr.Op == hclsyntax.OpLogicalAnd || binaryOpExpr.Op == hclsyntax.OpLogicalOr {
+				// Check if left side is a null check
+				if isNullCheck(binaryOpExpr.LHS) {
+					// Check if right side references the same variable as the left side
+					if referencesNullCheckedVar(binaryOpExpr.LHS, binaryOpExpr.RHS) {
+						if err := runner.EmitIssue(
+							r,
+							"Short-circuit evaluation is not supported in Terraform. Use a conditional expression (condition ? true : false) instead.",
+							binaryOpExpr.Range(),
+						); err != nil {
+							return hcl.Diagnostics{
+								{
+									Severity: hcl.DiagError,
+									Summary:  "Failed to emit issue",
+									Detail:   err.Error(),
+								},
+							}
+						}
+					}
+				}
+			}
+		}
+		return nil
+	}))
+
+	if diags.HasErrors() {
+		return diags
+	}
+	return nil
+}
+
+// isNullCheck determines if an expression is checking for null
+func isNullCheck(expr hcl.Expression) bool {
+	if binaryOpExpr, ok := expr.(*hclsyntax.BinaryOpExpr); ok {
+		if binaryOpExpr.Op == hclsyntax.OpEqual || binaryOpExpr.Op == hclsyntax.OpNotEqual {
```

Comment:
Strictly speaking, the only valid combinations are `OpLogicalAnd` with `OpNotEqual` and `OpLogicalOr` with `OpEqual`. However, in practice, an expression like `var.obj == null && var.obj.enabled` is rarely useful, so this may be sufficient as it is.